qsimcirq/qsim_simulator.py

```diff
@@ -62,6 +64,124 @@ def __init__(self, qsim_options: dict = {}):
+        results. Measurement results are stored in a 2-dimensional
+        numpy array, the first dimension corresponding to the repetition
+        and the second to the actual boolean measurement results (ordered
+        by the qubits being measured.)
+
+    Raises:
+        NotImplementedError: If there are non-terminal measurements in the
+            circuit.
+        ValueError: If there are multiple MeasurementGates with the same key,
+            or if repetitions is negative.
+    """
+    if not isinstance(program, qsimc.QSimCircuit):
+      program = qsimc.QSimCircuit(program, device=program.device)
+
+    if not program.are_all_measurements_terminal():
+      raise NotImplementedError("support for non-terminal measurement is not yet implemented")
+
+    # Compute indices of measured qubits
+    ordered_qubits = ops.QubitOrder.DEFAULT.order_for(program.all_qubits())
+    ordered_qubits = list(reversed(ordered_qubits))
+
+    qubit_map = {
+      qubit: index for index, qubit in enumerate(ordered_qubits)
+    }
+
+    # Computes
+    # - the list of qubits to be measured
+    # - the start (inclusive) and end (exclusive) indices of each measurement
+    # - a mapping from measurement key to measurement gate
+    measurement_ops = [
+      op for _, op, _ in program.findall_operations_with_gate_type(ops.MeasurementGate)
+    ]
+    measured_qubits = []  # type: List[ops.Qid]
+    bounds = {}  # type: Dict[str, Tuple]
+    meas_ops = {}  # type: Dict[str, cirq.MeasurementGate]
+    current_index = 0
+    for op in measurement_ops:
+      gate = op.gate
+      key = protocols.measurement_key(gate)
+      meas_ops[key] = gate
+      if key in bounds:
+        raise ValueError("Duplicate MeasurementGate with key {}".format(key))
+      bounds[key] = (current_index, current_index + len(op.qubits))
+      measured_qubits.extend(op.qubits)
+      current_index += len(op.qubits)
+
+    indices = [qubit_map[qubit] for qubit in measured_qubits]
+
+    # Want to check all bitstrings
+    n_qubits = len(program.all_qubits())
+    bitstrings = [i for i in range(2**n_qubits)]
+    bitstrings = [format(bs, 'b').zfill(n_qubits) for bs in bitstrings]
+
+    # Set qsim options
+    options = {}
+    options.update(self.qsim_options)
+    options['c'] = program.translate_cirq_to_qsim(ops.QubitOrder.DEFAULT)
+
+    results = {}
+    for key, bound in bounds.items():
+      boundlen = bound[1]-bound[0]
+      results[key] = np.ndarray(shape=(repetitions, bound[1]-bound[0]), dtype=int)
+
+    for i in range(repetitions):
+      qsim_state = qsim.qsim_simulate_fullstate(options)
+      amplitudes = QSimSimulatorState(qsim_state, qubit_map).state_vector
+
+      # Convert amplitudes to probabilities
+      probabilities = np.abs(amplitudes)**2
+      probabilities /= probabilities.sum()
+
+      # format the bitstring in terms of measurement gates
+      result = np.random.choice(bitstrings, p=probabilities)
```

Comment:
Given that the state vector object has O(2^N) elements for N qubits, we want to avoid copying it or operating on it in Python. This will require a separate pybind method which extracts the results from a simulation.

Incidentally, measurements in qsim already calculate their results - we just don't store them. Preferably we would use these results in the pybind method.

Reply:
PR #162 added the `qsim_sample` method, which returns a flat list of measurement values ordered by (time, qubit index). Calling this instead of `qsim_simulate_fullstate` should allow you to skip the measurement-extraction steps here (although converting to a dict will still be necessary).

Note that `qsim_sample` works for both terminal and non-terminal measurements, but must be called separately for each repetition since measurements are nondeterministic. For circuits with only terminal measurements, calling `simulate()` once and extracting results from the resulting state vector is likely faster.